src/main/controllers/ChallengeController.kt

```diff
@@ -24,37 +43,44 @@ class ChallengeController: DefaultController<Challenge>(), RestController<Challe
 
-    @Throws(NotFoundException::class)
     override fun findOne(user: UserAccount, id: Int): SOAResult<Challenge> {
-        val findChallengeResult = DaoService.execute {
-            Challenge.findById(id)
-        }
+        val challenge = findChallengeById(id)
 
-        if (findChallengeResult.result != SOAResultType.SUCCESS || findChallengeResult.data == null) {
-            throw NotFoundException("Challenge not found")
-        }
-
-        return SOAResult(findChallengeResult.result, findChallengeResult.message, findChallengeResult.data!!)
+        return SOAResult(SOAResultType.SUCCESS, "", challenge)
     }
 
     fun complete(user: UserAccount, request: Request): SOAResult<Challenge> {
         throw NotImplementedError()
     }
 
-    fun share(user: UserAccount, request: Request): SOAResult<TransactionWithNewUser> {
-        return DaoService.execute {
+    @Throws(ForbiddenException::class)
+    fun share(user: UserAccount, request: Request): SOAResult<TransactionWithNewUser?> {
+        val finalResult = SOAResult<TransactionWithNewUser?>(SOAResultType.FAILURE, null, null)
+
+        val challengeId = request.input["challengeId"] as Int
+        val publicKeyToShareWith = request.input["publicKeyToShareWith"] as String?
+        val shares = request.input["shares"] as Int
+        val expiration = request.input["expiration"] as String?
+        val emailToShareWith = request.input["emailToShareWith"] as String?
+
+        val validateApiKeyResult = DaoService.execute {
             ValidateApiKeyService.execute(user, request.input["secretKey"] as String)
+        }
+        DaoService.throwOrReturn(validateApiKeyResult.result, validateApiKeyResult.message)
+
+        val challenge = findChallengeById(challengeId)
 
-            val challengeId = request.input["challengeId"] as Int
-            val challenge = Challenge.findById(challengeId)!!
-            val publicKeyToShareWith = request.input["publicKeyToShareWith"] as String?
-            val shares = request.input["shares"] as Int
-            val expiration = request.input["expiration"] as String?
-            val emailToShareWith = request.input["emailToShareWith"] as String?
+        DaoService.execute {
+            val serviceResult = ShareChallengeService.execute(user, challenge, shares, publicKeyToShareWith, emailToShareWith, expiration)
+            finalResult.result = serviceResult.result
+            finalResult.message = serviceResult.message
+            finalResult.data = serviceResult.data
+        }
 
-            val shareChallengeResult = ShareChallengeService.execute(user, challenge, shares, publicKeyToShareWith, emailToShareWith, expiration)
-            DaoService.throwOrReturn(shareChallengeResult.result, shareChallengeResult.message)
-            return@execute shareChallengeResult.data!!
+        if (finalResult.result != SOAResultType.SUCCESS && finalResult.message != null) {
+            throw ForbiddenException(finalResult.message as String)
```

Comment:
im not sure i totally understand this part...so any error that happens in the DAOService.execute is going to be treated as a forbidden exception? what is theres a db connection error? or any number of other errors?